### ISSUE TITLE
Update extract_1d to handle ModelContainer as input

### ISSUE THREAD

**Issue by hbushouse** [STScI-JWST/jwst#1264]
The cube_build step was recently updated to return results in the form of a ModelContainer, even when a single IFU cube is being built. When that result is then fed to the extract_1d step in either calspec2 or calspec3, extract_1d responds with the warning:

2017-10-02 09:36:34,778 - stpipe.Spec2Pipeline.cube_build - INFO - Step cube_build done
2017-10-02 09:36:36,140 - stpipe.Spec2Pipeline.extract_1d - INFO - Prefetch for EXTRACT1D reference file is '/grp/crds/cache/references/jwst/jwst_miri_extract1d_0003.json'.
2017-10-02 09:36:36,369 - stpipe.Spec2Pipeline.extract_1d - INFO - Step extract_1d running with args (<jwst.datamodels.container.ModelContainer object at 0x7fe854191cd0>,).
2017-10-02 09:36:36,408 - stpipe.Spec2Pipeline.extract_1d - WARNING - Input is a <class 'jwst.datamodels.container.ModelContainer'>,
2017-10-02 09:36:36,408 - stpipe.Spec2Pipeline.extract_1d - WARNING - which was not expected for extract_1d.

and then errors and tracebacks occur.

The extract_1d step needs to be updated to allow for a ModelContainer as input and loop over each model in the container (many times there will only be a single model in the container).

**Comment by hbushouse** [comment]
Fixed by #1271.

**Comment by hbushouse** [comment]
We still seem to have one remaining issue with this situation, which is that only 1 output x1d product is getting saved. When presented with a ModelContainer as input, extract_1d is properly looping over each model in the container and appears to be correctly creating multiple results, also in the form of a ModelContainer, containing multiple MultiSpecModel's. But when the step saves results, only 1 file is getting created, containing only 1 result (not sure if the data are from the first or second spectrum). So there's some kind of issue yet with the step infrastructure saving a ModelContainer.

Here's the log of an example run:
```
2017-10-20 09:59:15,712 - stpipe.Spec3Pipeline.cube_build - INFO - The desired cubes covers the MIRI Channels: ['1', '2']
2017-10-20 09:59:15,712 - stpipe.Spec3Pipeline.cube_build - INFO - The desired cubes covers the MIRI subchannels: ['SHORT', 'SHORT']
2017-10-20 09:59:15,712 - stpipe.Spec3Pipeline.cube_build - INFO - Reading  cube parameter file /grp/crds/cache/references/jwst/jwst_miri_cubepar_0001.fits
2017-10-20 09:59:15,828 - stpipe.Spec3Pipeline.cube_build - INFO - Output Cubes are single channel, single sub-channel IFU Cubes
2017-10-20 09:59:15,828 - stpipe.Spec3Pipeline.cube_build - INFO - Number of IFUCubes produced by a this run 2
2017-10-20 09:59:15,899 - stpipe.Spec3Pipeline.cube_build - INFO - Region of interest  0.100000 0.002500
2017-10-20 09:59:15,899 - stpipe.Spec3Pipeline.cube_build - INFO - Number of bands in cube  1
2017-10-20 09:59:15,899 - stpipe.Spec3Pipeline.cube_build - DEBUG - Working on data  from 1,SHORT
2017-10-20 09:59:15,899 - stpipe.Spec3Pipeline.cube_build - DEBUG - number of files 4 
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Cube Geometry:
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - axis# Naxis  CRPIX    CRVAL      CDELT(arc sec)  MIN & Max (xi,eta arc sec)
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 1    52  27.00  45.00000739   0.13000000  -3.44499987   3.44499987
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 2    38  20.00   0.00016545   0.13000000  -2.53499991   2.53499991
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 3   372   1.00   4.84711374   0.00250000   4.84711374   5.77711372
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Cube covers channel, subchannel: 1 SHORT 
2017-10-20 09:59:20,926 - stpipe.Spec3Pipeline.cube_build - INFO - Output Name Test_4dither_ch1-short_s3d.fits
...
2017-10-20 10:01:48,782 - stpipe.Spec3Pipeline.cube_build - DEBUG - Working on data  from 2,SHORT
2017-10-20 10:01:48,782 - stpipe.Spec3Pipeline.cube_build - DEBUG - number of files 4 
2017-10-20 10:01:53,720 - stpipe.Spec3Pipeline.cube_build - INFO - Cube Geometry:
2017-10-20 10:01:53,720 - stpipe.Spec3Pipeline.cube_build - INFO - axis# Naxis  CRPIX    CRVAL      CDELT(arc sec)  MIN & Max (xi,eta arc sec)
2017-10-20 10:01:53,720 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 1    60  31.00  44.99999116   0.13000000  -3.96499985   3.96499985
2017-10-20 10:01:53,720 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 2    46  24.00   0.00008002   0.13000000  -3.05499989   3.05499989
2017-10-20 10:01:53,720 - stpipe.Spec3Pipeline.cube_build - INFO - Axis 3   358   1.00   7.44517473   0.00400000   7.44517473   8.87717480
2017-10-20 10:01:53,721 - stpipe.Spec3Pipeline.cube_build - INFO - Cube covers channel, subchannel: 2 SHORT 
2017-10-20 10:01:53,721 - stpipe.Spec3Pipeline.cube_build - INFO - Output Name Test_4dither_ch2-short_s3d.fits
...
2017-10-20 10:04:31,428 - stpipe.Spec3Pipeline.cube_build - INFO - Step cube_build done
2017-10-20 10:04:34,430 - stpipe.Spec3Pipeline.extract_1d - INFO - Prefetch for EXTRACT1D reference file is '/grp/crds/cache/references/jwst/jwst_miri_extract1d_0003.json'.
2017-10-20 10:04:35,925 - stpipe.Spec3Pipeline.extract_1d - INFO - Prefetch for EXTRACT1D reference file is '/grp/crds/cache/references/jwst/jwst_miri_extract1d_0003.json'.
2017-10-20 10:04:36,607 - stpipe.Spec3Pipeline.extract_1d - INFO - Step extract_1d running with args (<jwst.datamodels.container.ModelContainer object at 0x7fdb710a2050>,).
2017-10-20 10:04:36,746 - stpipe.Spec3Pipeline.extract_1d - DEBUG - Input is a ModelContainer
2017-10-20 10:04:36,746 - stpipe.Spec3Pipeline.extract_1d - DEBUG - Input contains 2 items
2017-10-20 10:04:38,968 - stpipe.Spec3Pipeline.extract_1d - INFO - Using EXTRACT1D reference file /grp/crds/cache/references/jwst/jwst_miri_extract1d_0003.json
2017-10-20 10:04:39,129 - stpipe.Spec3Pipeline.extract_1d - DEBUG - slitname=MIR_MRS
2017-10-20 10:04:39,129 - stpipe.Spec3Pipeline.extract_1d - INFO - source_type = extended
2017-10-20 10:04:39,286 - stpipe.Spec3Pipeline.extract_1d - DEBUG - IFU 1-D extraction parameters:
2017-10-20 10:04:39,286 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   x_center = 26.0
2017-10-20 10:04:39,286 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   y_center = 19.0
2017-10-20 10:04:39,286 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   width = 19.0
2017-10-20 10:04:39,286 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   height = 19.0
2017-10-20 10:04:39,287 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   theta = 0.0 degrees
2017-10-20 10:04:39,287 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   subtract_background = False
2017-10-20 10:04:39,287 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   method = subpixel
2017-10-20 10:04:39,287 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   subpixels = 10
2017-10-20 10:04:39,975 - stpipe.Spec3Pipeline.extract_1d - INFO - Data have been flux calibrated; setting net to 0.
2017-10-20 10:04:42,717 - stpipe.Spec3Pipeline.extract_1d - INFO - Using EXTRACT1D reference file /grp/crds/cache/references/jwst/jwst_miri_extract1d_0003.json
2017-10-20 10:04:42,875 - stpipe.Spec3Pipeline.extract_1d - DEBUG - slitname=MIR_MRS
2017-10-20 10:04:42,875 - stpipe.Spec3Pipeline.extract_1d - INFO - source_type = extended
2017-10-20 10:04:43,032 - stpipe.Spec3Pipeline.extract_1d - DEBUG - IFU 1-D extraction parameters:
2017-10-20 10:04:43,032 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   x_center = 30.0
2017-10-20 10:04:43,032 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   y_center = 23.0
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   width = 23.0
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   height = 23.0
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   theta = 0.0 degrees
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   subtract_background = False
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   method = subpixel
2017-10-20 10:04:43,033 - stpipe.Spec3Pipeline.extract_1d - DEBUG -   subpixels = 10
2017-10-20 10:04:43,712 - stpipe.Spec3Pipeline.extract_1d - INFO - Data have been flux calibrated; setting net to 0.
2017-10-20 10:04:44,074 - stpipe.Spec3Pipeline.extract_1d - INFO - Saving file Test_4dither_x1d.fits
2017-10-20 10:04:44,522 - stpipe.Spec3Pipeline.extract_1d - INFO - Step extract_1d done
2017-10-20 10:04:44,523 - stpipe.Spec3Pipeline - INFO - Ending calwebb_spec3
2017-10-20 10:04:44,525 - stpipe.Spec3Pipeline - INFO - Step Spec3Pipeline done
```

Note that while 2 products were saved by cube_build - Test_4dither_ch1-short_s3d.fits and Test_4dither_ch2-short_s3d.fits - there's only one product saved by extract_1d - Test_4dither_x1d.fits - and it only contains one spectral table. There should be two separate x1d products, each corresponding to the cube_build products.

Example data are in /grp/jwst/ssb/bushouse/jwst_data/MIRI/MRS/. The above test was done using:
```
strun calwebb_spec3.cfg cube_build_4dither_495_asn.json --steps.mrs_imatch.skip=true -v
```

(mrs_imatch was skipped just to save time - it works fine on these data)

**Comment by hbushouse** [comment]
Changing milestone to B7.2, because technically we don't need to provide support for level-3 pipelines until B7.2, but if we can get this fixed in time to include in one more potential B7.1 patch, that would at least allow users/testers to run the calspec3 pipeline properly on IFU data.

**Comment by stscieisenhamer** [comment]
@hbushouse This data does not seem to run anymore. Getting the following:

```
Traceback (most recent call last):
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/stpipe/cmdline.py", line 297, in step_from_cmdline
    step.run(*positional)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/stpipe/step.py", line 405, in run
    step_result = self.process(*args)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/pipeline/calwebb_spec3.py", line 117, in process
    result = self.mrs_imatch(result)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/stpipe/step.py", line 405, in run
    step_result = self.process(*args)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/mrs_imatch/mrs_imatch_step.py", line 92, in process
    degree=degree)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/mrs_imatch/mrs_imatch_step.py", line 266, in _match_models
    cube_models = cbs.process(models)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/cube_build/cube_build_step.py", line 272, in process
    thiscube.setup_ifucube_wcs()
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/cube_build/ifu_cube.py", line 228, in setup_ifucube_wcs
    cube_build_wcs_util.setup_wcs(self)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/cube_build/cube_build_wcs_util.py", line 118, in setup_wcs
    self.instrument_info)
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/cube_build/cube_build_wcs_util.py", line 308, in find_footprint_MIRI
    detector2v23 = input.meta.wcs.get_transform('detector', 'v2v3')
  File "/Users/eisenham/Documents/ssbdev/jwst/jwst/datamodels/properties.py", line 207, in __getattr__
    raise AttributeError("No attribute '{0}'".format(attr))
AttributeError: No attribute 'wcs'
```

**Comment by hbushouse** [comment]
Yes, same error here when trying to test this case, as well as #1288. If I feed an ASN file to just the cube_build step, it works fine. But if I feed the same input to calwebb_spec3, it always dies at this same line. I wonder if the wcs object is not getting copied properly when the ModelContainer is setup? Because if I just load a single input file into a datamodel, the wcs object is there and is populated.

**Comment by hbushouse** [comment]
Problem solved. The datamodels were modified recently such that 2D IFU images, like these, need DATAMODL='IFUImageModel' in their header, instead of the old 'ImageModel'. Update the DATAMODL keyword value and it should run.

**Comment by hbushouse** [comment]
All problems here have now been fixed. A 3-exposure MIRI MRS ASN goes through calwebb_spec3 and produces separate s3d products for each channel, as well as two separate x1d products, one for each channel. Likely fixed by all the file handling updates in #1316.
